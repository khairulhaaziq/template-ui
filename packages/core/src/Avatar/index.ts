export {
  injectAvatarRootContext,
  default as AvatarRoot,
  type AvatarRootProps,
} from './AvatarRoot.vue'
export {
  default as AvatarFallback,
  type AvatarFallbackProps,
} from './AvatarFallback.vue'
export {
  default as AvatarImage,
  type AvatarImageEmits,
  type AvatarImageProps,
} from './AvatarImage.vue'
