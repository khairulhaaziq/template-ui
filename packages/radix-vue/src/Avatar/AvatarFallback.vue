<template><div class="flex h-full w-full items-center justify-center rounded-full bg-muted"><slot /></div></template>