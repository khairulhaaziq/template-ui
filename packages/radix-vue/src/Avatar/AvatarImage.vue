<template>
	<img :src="props.src" :alt="props.alt">
</template>

<script setup>
const props = defineProps({
	src: String,
	alt: String
})
</script>