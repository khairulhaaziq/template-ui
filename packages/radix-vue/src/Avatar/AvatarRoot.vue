<script lang="ts">
import type { Ref } from 'vue'
import type { ImageLoadingStatus } from './utils'
import { createContext } from '@/shared'

export interface AvatarRootProps extends PrimitiveProps {}

export type AvatarRootContext = {
  imageLoadingStatus: Ref<ImageLoadingStatus>
}

export const [injectAvatarRootContext, provideAvatarRootContext]
  = createContext<AvatarRootContext>('AvatarRoot')
</script>

<script setup lang="ts">
import { ref } from 'vue'
import { Primitive, type PrimitiveProps } from '../Primitive'

const props = withDefaults(defineProps<AvatarRootProps>(), {
  as: 'span',
})

provideAvatarRootContext({
  imageLoadingStatus: ref<ImageLoadingStatus>('loading'),
})
</script>

<template>
  <Primitive :as="as" :as-child="props.asChild">
    <slot />
  </Primitive>
</template>
