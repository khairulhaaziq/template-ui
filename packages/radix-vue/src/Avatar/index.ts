export { default as AvatarRoot } from "./AvatarRoot.vue";
export { default as AvatarImage } from "./AvatarImage.vue";
export { default as AvatarFallback } from "./AvatarFallback.vue";
