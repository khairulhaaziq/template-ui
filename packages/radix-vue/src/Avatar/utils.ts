import { onMounted, onUnmounted, ref } from 'vue'

export type ImageLoadingStatus = 'idle' | 'loading' | 'loaded' | 'error'

export function useImageLoadingStatus(src?: string) {
  const loadingStatus = ref<ImageLoadingStatus>('idle')
  const isMounted = ref(false)
  onMounted(() => {
    if (!src) {
      loadingStatus.value = 'error'
      return
    }

    isMounted.value = true
    const image = new window.Image()

    const updateStatus = (status: ImageLoadingStatus) => () => {
      if (!isMounted.value)
        return
      loadingStatus.value = status
    }

    loadingStatus.value = 'loading'
    image.onload = updateStatus('loaded')
    image.onerror = updateStatus('error')
    image.src = src
  })

  onUnmounted(() => {
    isMounted.value = false
  })

  return loadingStatus
}
