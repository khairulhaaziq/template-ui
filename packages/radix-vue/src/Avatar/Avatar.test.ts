import { beforeAll, describe, expect, it } from 'vitest'
import { axe } from 'vitest-axe'
import Avatar from './story/_Avatar.vue'
import type { VueWrapper } from '@vue/test-utils'
import { mount } from '@vue/test-utils'
import { nextTick } from 'vue'
import { findByAltText, findByRole, findByText } from '@testing-library/vue'

const FALLBACK = 'CT'
const DELAY = 350

it('should pass axe accessibility tests', async () => {
  const wrapper = mount(Avatar)
  expect(await axe(wrapper.element)).toHaveNoViolations()
})

describe('given an Avatar with fallback and a working image', async () => {
  let wrapper: VueWrapper<InstanceType<typeof Avatar>>
  beforeAll(async () => {
    (window.Image as any) = class MockImage {
      onload: () => void = () => {}
      src = ''
      constructor() {
        setTimeout(() => {
          this.onload()
        }, DELAY)
        return this
      }
    }
    wrapper = mount(Avatar)
  })

  it('should render the fallback initially', async () => {
    await nextTick()
    expect(wrapper.html()).includes(FALLBACK)
  })

  it('should not render the image initially', () => {
    const image = wrapper.find('img')
    expect(image.exists()).toBeFalsy()
  })

  it('should render the image after it has loaded', async () => {
    const image = await findByRole(wrapper.element as HTMLElement, 'img')

    expect(image).toBeTruthy()
  })

  it('should have alt text on the image', async () => {
    const image = await findByAltText(wrapper.element as HTMLElement, 'Colm Tuite')
    expect(image).toBeTruthy()
  })

  it('should render as snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('given an Avatar with fallback and delayed render', () => {
  let wrapper: VueWrapper<InstanceType<typeof Avatar>>

  beforeAll(async () => {
    (window.Image as any) = class MockImage {
      onload: () => void = () => {}
      src = ''
      constructor() {
        setTimeout(() => {
          this.onload()
        }, DELAY)
        return this
      }
    }
    wrapper = mount(Avatar, {
      props: {
        delay: 300,
      },
    })
  })

  it('should not render a fallback immediately', async () => {
    expect(wrapper.text()).not.toContain(FALLBACK)
  })

  it('should render a fallback after the delay', async () => {
    expect(wrapper.text()).not.toContain(FALLBACK)
    const fallback = await findByText(wrapper.element as HTMLElement, FALLBACK)
    expect(wrapper.element).toContain(fallback)
  })
})
