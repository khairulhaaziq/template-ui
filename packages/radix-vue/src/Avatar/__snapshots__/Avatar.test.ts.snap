// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`given an Avatar with fallback and a working image > should render as snapshot 1`] = `"<span class="bg-blackA3 inline-flex h-[45px] w-[45px] select-none items-center justify-center overflow-hidden rounded-full align-middle"><img role="img" src="https://images.unsplash.com/photo-1492633423870-43d1cd2775eb?&amp;w=128&amp;h=128&amp;dpr=2&amp;q=80" class="h-full w-full rounded-[inherit] object-cover" alt="Colm Tuite"><!--v-if--></span>"`;
